software/runtime/kernel/synth.h

```diff
@@ -142,13 +461,340 @@ static inline void power_profile(int32_t *const a) {
       "bne %[idx], %[im_0], 3b \n\t"
+      "addi %[idx], %[idx], %[num_loops] \n\t"
+      ".balign 16 \n\t"
+      "4: \n\t"
+      "p.mac %[op_0],%[op_a0],%[op_a4] \n\t"
+      "p.mac %[op_1],%[op_a1],%[op_a5] \n\t"
+      "p.mac %[op_2],%[op_a2],%[op_a6] \n\t"
+      "p.mac %[op_3],%[op_a3],%[op_a7] \n\t"
+      "p.mac %[op_0],%[op_a0],%[op_a4] \n\t"
+      "p.mac %[op_1],%[op_a1],%[op_a5] \n\t"
+      "p.mac %[op_2],%[op_a2],%[op_a6] \n\t"
+      "p.mac %[op_3],%[op_a3],%[op_a7] \n\t"
+      "p.mac %[op_0],%[op_a0],%[op_a4] \n\t"
+      "p.mac %[op_1],%[op_a1],%[op_a5] \n\t"
+      "p.mac %[op_2],%[op_a2],%[op_a6] \n\t"
+      "p.mac %[op_3],%[op_a3],%[op_a7] \n\t"
+      "p.mac %[op_0],%[op_a0],%[op_a4] \n\t"
+      "p.mac %[op_1],%[op_a1],%[op_a5] \n\t"
+      "p.mac %[op_2],%[op_a2],%[op_a6] \n\t"
+      "p.mac %[op_3],%[op_a3],%[op_a7] \n\t"
+      "p.mac %[op_0],%[op_a0],%[op_a4] \n\t"
+      "p.mac %[op_1],%[op_a1],%[op_a5] \n\t"
+      "p.mac %[op_2],%[op_a2],%[op_a6] \n\t"
+      "p.mac %[op_3],%[op_a3],%[op_a7] \n\t"
+      "p.mac %[op_0],%[op_a0],%[op_a4] \n\t"
+      "p.mac %[op_1],%[op_a1],%[op_a5] \n\t"
+      "p.mac %[op_2],%[op_a2],%[op_a6] \n\t"
+      "p.mac %[op_3],%[op_a3],%[op_a7] \n\t"
+      "p.mac %[op_0],%[op_a0],%[op_a4] \n\t"
+      "p.mac %[op_1],%[op_a1],%[op_a5] \n\t"
+      "p.mac %[op_2],%[op_a2],%[op_a6] \n\t"
+      "p.mac %[op_3],%[op_a3],%[op_a7] \n\t"
+      "p.mac %[op_0],%[op_a0],%[op_a4] \n\t"
+      "p.mac %[op_1],%[op_a1],%[op_a5] \n\t"
+      "p.mac %[op_2],%[op_a2],%[op_a6] \n\t"
+      "p.mac %[op_3],%[op_a3],%[op_a7] \n\t"
+      "addi %[idx], %[idx], -1 \n\t"
+      "bne %[idx], %[im_0], 4b \n\t"
+      "addi %[idx], %[idx], %[num_loops] \n\t"
       : [op_0] "=&r"(op_0), [op_1] "=&r"(op_1), [op_2] "=&r"(op_2),
         [op_3] "=&r"(op_3), [a0] "=&r"(a0), [a1] "=&r"(a1), [a2] "=&r"(a2),
         [a3] "=&r"(a3), [a4] "=&r"(a4), [a5] "=&r"(a5), [a6] "=&r"(a6),
         [a7] "=&r"(a7), [idx] "+&r"(idx)
       : [im_0] "r"(im_0), [num_loops] "I"(num_loops), [op_a0] "r"(op_a0),
         [op_a1] "r"(op_a1), [op_a2] "r"(op_a2), [op_a3] "r"(op_a3),
         [op_a4] "r"(op_a4), [op_a5] "r"(op_a5), [op_a6] "r"(op_a6),
-        [op_a7] "r"(op_a7), [a] "r"(a)
+        [op_a7] "r"(op_a7), [a_local] "r"(addr_local)
+      : "memory");
+}
+
+static inline void power_profile_terapool_lw_rtile(int32_t *const a,
+                                                   uint32_t remote_tile) {
+
+  // Address to load from are [addr:addr+3]
+  volatile int32_t *addr_remote_tile = &a[remote_tile];
+
+  // Do this loop M times
+  int const num_loops = 2;
+  int32_t idx = num_loops;
+  int32_t im_0 = 0;
+  // Temporary registers
+  register int32_t a0;
+  register int32_t a1;
+  register int32_t a2;
+  register int32_t a3;
+  register int32_t a4;
+  register int32_t a5;
+  register int32_t a6;
+  register int32_t a7;
+  __asm__ volatile(
+      // NOP to wait for instruction cache refill
```

Comment:
Why do we have to do so many nops? Won't they pollute the I$, and we should just execute the loop multiple times?